### DISCUSSION TITLE
Grand resource usage in Docker environment

### DISCUSSION THREAD

**Discussion by Jurrer** [gramps-project/gramps-web#315] [General]
Hi
I am from-time-to-time user. My tree is rather small and incomplete.

For a web app with api, Grampsweb is consuming enormous resources.
I'd like to enquiry,
How are the docker images built? Is it possible to use a lighter alpine image perhaps?
I couldn't dig a source of mysterious image `FROM dmstraub/gramps:5.1.6` too.

As for my deduction techniques, 
I found out by sorting `/var/lib/docker/overlay2` folder, where two biggest directories belonged to Gramps 
(among ~50 containers)
```
root@host:/var/lib/docker/overlay2# du -hs * | sort -rh | head -2
2.1G	ed58b6c14946392bfbad5985751544fb226d2622b0baccb63e0819bc2ea5f538
2.1G	29e390349624261964f1ddc1ce34988636ab7267b343c393da5d4298a7543719
```
That's ~4.2G!

On another note, mounted volumes didn't take much space.

BUT

Much worse revelation was finding out that idling app is consuming >1.5G of RAM!

Don't take me the wrong way. I love the app. Love the idea. Love to connect with my family.

I'd just like to open a dialog in regards to optimizations.
Off the top of my head maybe support for external postgres DB?
More modular approach?
Let me know what you think.

**Comment by DavidMStraub** [comment]
Hi,

- Source of the "mysterious" image: https://github.com/DavidMStraub/gramps-docker, https://hub.docker.com/r/dmstraub/gramps
- Resource usage: https://www.grampsweb.org/cpu-limited/ (that also applies to RAM, each gunicorn worker consumes a fixed amount even when idle)
- Postgres: https://www.grampsweb.org/Postgres/ but note that this will **not reduce resource usage** and not speed up your app
- Large docker blobs: not much to say unless you specify what that is. Could be simply the debian base image layer.

Most of the issues you describe have nothing to do with Gramps Web, but with Docker or Gunicorn.

Converting to discussion.

**Comment by nothing2obvi** [comment]
I'm wondering the same thing....Out of all of my apps, even without limits placed on resources, Grampsweb takes up the most resources, which is very surprising.

**Reply by filippogrande** [reply]
I also have noticed the same behavour, maybe it could be put as a new feature in the issuepage, when the developers finish most of the feature that they want or the community wants to be implemented. The feature could be something like: 
Improuved resource usage, diminuing ram requirments for running the immage.

**Comment by Citrus0299** [comment]
@umlaeute continue here. @DavidMStraub if you don't care about an issue just unsubscribe and let people discuss it instead of locking it. Hope you just had a bad day and this is not how things are going around here.

**Reply by DavidMStraub** [reply]
I can tell you why I locked it. The issue was closed because I made a decision, namely that a "light" image is out of scope for the https://github.com/gramps-project/gramps-web-api or https://github.com/gramps-project/gramps-web repositories, so any further discussion on non-standard images is off-topic for that issue.

A discussion whether somebody wants to start and maintain a light image is more than welcome, but that issue was not the appropriate place for it.

>  if you don't care about an issue ... Hope you just had a bad day

I find this comment discrespectful and as such against the sprit of the [code of conduct](https://github.com/gramps-project/gramps-web/blob/main/CODE_OF_CONDUCT.md).

**Reply by Citrus0299** [reply]
The discussion was about fixing the existing image and ci build which can be done btw. The web api image is also completely bloated. Why maintain two images that's completely unnecessary if the existing one has so much room for improvement?
What is the better place for a discussion like that if not the repo where the images are built?

I honestly believe that you don't care about the issue, otherwise you would have linked existing discussions or issues and the correct place for a discussion. You just instantly closed the issue with a vague explanation about some existing issue which can be found god knows where. After another user pointed out possible improvements you locked the issue.

You getting offended by two messages and instantly pulling out the CoC for facing criticism proves me even more right.
There is no such thing as being against the spirit of the CoC. Even if a CoC is as vague as yours and could apply to any form of criticism whatsoever. If you feel like there is a true violation, you are on the board anyway, go ahead and issue a formal warning.

**Reply by DavidMStraub** [reply]
Not caring and defining something as out of scope are two different things.

If you were to acknowledge how much work I am investing into this project, you would understand that I have to be very mindful of scope creep if I want this project to make any progress at all. This applies in particular to packaged distributions (like the docker images). The current scope is that I'm providing an image, and documentation for it, that provides all features (including semantic search that leads to a large image size). Everything else needs to be provided by somebody outside of the core project.

Claiming that I don't care just because I am managing my precious time in the interest of the project's progress is what I call disrespectful.

**Reply by DavidMStraub** [reply]
For other people's reference, the issue being referred to is this: https://github.com/gramps-project/gramps-web/issues/695

**Reply by emyoulation** [reply]
I think I see David's concern. And Citrus' point.

When David first announced Gramps Web on the Discourse forum, I jumped in  in the discussion. Although I had no need   to start managing a collaborative genealogy project, it needed a strong community to succeed. Several attempts in that vein had withered on the vine. So I tried to help build that community.

David has been incredibly energenic and dedicated. But he is stretched too far. Frankly, I don't know how he does so much and am a huge admirer. He has been architect and developer, documentation and support, marketing, quality control, and production. 

He is starting to get some people coding contributions. But he cannot expand into oversight of new development directions. And optimization of the delivery system is a huge project expansion.

This idea is great but it needs to very distinctly NOT divert David. It is too likely to intrigue him if the development discussion takes place in an area that he administers. So how about forking the repository and starting an independent discussion area here on GitHub and start a new thread in the "Development" category of Discourse too?

**Reply by LoganJFisher** [reply]
> I can tell you why I locked it. The issue was closed because I made a decision, namely that a "light" image is out of scope for the https://github.com/gramps-project/gramps-web-api or https://github.com/gramps-project/gramps-web repositories, so any further discussion on non-standard images is off-topic for that issue.

Hello David. First of all, thank you for all your work on this. I'm enjoying using Gramps Web (although have experience a minor bug I'm sure you're already aware of, wherein dropdown menus sometimes appear blank until the page is refreshed). The RAM consumption is unfortunately a bit of an issue for me as I'm running on an RPi, but I've ultimately resolved to mitigate it by just not running this when I'm not actively using it — it's not something I currently need to leave running 24/7. That being said, I'd definitely support RAM optimization (perhaps utilizing SWAP as necessary to help reduce this to some degree). I'm not really understanding your above reasoning though — why would this necessitate the development of a new "light" image instead of being the pursuit of RAM optimization for the existing images? Please excuse my naivety — this sort of development is well outside of my wheelhouse.

**Reply by DavidMStraub** [reply]
Hi @LoganJFisher, actually - unless I am misunderstanding something - this discussion is not related at all to RAM consumption, but to the size of the Docker image.

For RAM consumption, you might find [this documentation article](https://www.grampsweb.org/install_setup/cpu-limited/) helpful.

**Reply by LoganJFisher** [reply]
@DavidMStraub I think the original topic was on a multitude of optimization matters, including the size of the Docker image, but they definitely also mentioned the idle RAM consumption. I'll take a look at that documentation — thanks.

**Reply by DavidMStraub** [reply]
You're right. But it makes sense to discuss them separately as they are unrelated.

**Reply by DavidMStraub** [reply]
(Note the link to the documentation was already given in my original answer.)

**Reply by LoganJFisher** [reply]
> You're right. But it makes sense to discuss them separately as they are unrelated.

Fair enough. Should I take that as prompting to open an issue specific to RAM optimization?

> (Note the link to the documentation was already given in my original answer.)

Apologies. I missed that.

**Reply by DavidMStraub** [reply]
> Fair enough. Should I take that as prompting to open an issue specific to RAM optimization?

You are free to do it, but please first try the various configuration options and also have a look at the gunicorn configuration (e.g. the `--preload` option). In the end, Gramps Web is a Flask app deployed with gunicorn, and a big share of the RAM usage is due to how these underlying libraries work, so apart from configuration, there is little that can be done on our end.

**Reply by LoganJFisher** [reply]
Is there any issue with dropping both workers to 1? Gramps Web still seems stable. The CPU usage fluctates from 0.1% to 25%, and the RAM seems to stick around 10.5%. Still a bit high on the high end, but far more tolerable than it was. Just want to make sure there won't be any stability issues from dropping them both down to 1.

**Reply by DavidMStraub** [reply]
I hope you understand if this is my last reply on this topic in *this* thread as I have asked repeatedly to discuss it separately.

If you only have 1 gunicorn worker, you cannot process requests simultaneously. If you have many users, this can be a problem; for a single user, not so much. But this is independent of Gramps Web, you can read more in the gunicorn documentation. Similarly for celery.
